Created a couchdb storage example using petsets.

James1912 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Its an example of a couchdb 2.0.0 deployment using petsets. 

**Special notes for your reviewer**:

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
```

James1912 committed:
Created a couchdb storage exsample using petsets.

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: in-tcp-port
+    port: 80
+    targetPort: 5984
+  - name: rsync-port
```
Remove this port, rsync not used in CouchDB

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: ssh-port
+    port: 22
+    targetPort: 22
+  - name: erlang-port
```
Maybe call this one "EPMD"?

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: erlang-port
+    port: 4369
+    targetPort: 4369
+  - name: foo-port
```
Rename this one to something Erlang-specific.

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+spec:
+  #clusterIP: 100.10.10.101
+  ports:
+  - name: in-tcp-port
```
Why are you exposing CouchDB on port 80 here?

James1912 replied:
This was so the users could hit the service `kubetest001.defult.svc.cluster.local`.  Did this because I have not bundled up HAproxy in this example.

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: rsync-port
+    port: 873
+    targetPort: 873
+  - name: ssh-port
```
Can you remove this?

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: admin-port
+    port: 5986
+    targetPort: 5986
+  - name: ssl-port
```
Would have expected 6984 for SSL in CouchDB-land ...

James1912 replied:
Silly me, its not even configured in the continer.  I have removed this port mapping

kocolosk commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+        name: couchdb
+        imagePullPolicy: Always
+        ports:
+        - containerPort: 873
```
Same comments about ports apply here. I would have expected to see just the ones needed for a CouchDB cluster.

kocolosk commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+    spec:
+      containers:
+      - env:
+        - name: POD_IP
```
Do you still need POD_IP in the land of Pet Sets? This felt like a bit of a hack when I added it before 1.3

James1912 replied:
I use the pod IP for getting the FQDN in the join_cluster script `socket.getfqdn(os.environ['POD_IP'])`

kocolosk replied:
Correct me if I'm wrong, but all you really need from that FQDN is the namespace and the name of the service to make the appropriate SRV query, right? I believe those are already available in the pod environment, I think using them directly would be cleaner than using the POD_IP to extract them from the FQDN.

MHBauer commented on examples/storage/couchdb/README.md:
````diff
@@ -0,0 +1,84 @@
+If your running minikube you can discover the cluster address as 
+follows: 
+```
+minikube service kubetest001 --url
````
What's an equivalent for those not running minikube?

James1912 replied:
A `kubectl get service couchdb001` will get the user the external IP of the service. Now it mentioned in the README

MHBauer commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: erlang-port
+    port: 4369
+    targetPort: 4369
+  - name: foo-port
```
Is this port needed? `foo` seems generic and a bad name to use. Is this an erlang or couch thing?

kocolosk replied:
Agree on the poor choice of name. The port itself is definitely needed, though - this is the port that the `vm.args` configures the Erlang VM to listen on for connection attempts from other Erlang nodes.

MHBauer commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+        pod.alpha.kubernetes.io/init-containers: '[
+            {
+            "name": "configer-couch",
+            "image": "glyndwr/kube-init-couchdb",
```
is there more info on this image? it doesn't seem to be an official couchdb image.

James1912 replied:
Sure you can see how this image is build in `examples/storage/couchdb/images/init_couch/`

MHBauer commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+        - containerPort: 4369
+          name: erlang-port
+        - containerPort: 9100
+          name: foo-port
```
again with foo port.

MHBauer commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+            {
+            "name": "configer-couch",
+            "image": "glyndwr/kube-init-couchdb",
+            "command": ["/bin/bash", "-c", "/./configer_couch.sh"],
```
what's up with `/./`?  Is there something special going on here?

James1912 replied:
So this is a hangover from when all of the kube specific configuration happened in the one container that included couchdb.  So the entrypoint was not set to `/`

MHBauer commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,114 @@
+            command:
+              - /bin/bash
+              - -c
+              - /./join_cluster.py _users
```
`/./` again.

MHBauer commented on examples/storage/couchdb/images/init_couch/configer_couch.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
name of this file?
configer -> configure

MHBauer commented on examples/storage/couchdb/couchdb-petset-headless-service.yaml:
```diff
@@ -0,0 +1,21 @@
+  labels:
+    tier: backend
+    app: couchdb
+  name: kubetest001-backend
```
I think it would be better to have a better name for these. `couchdb-backend` or `couchdb-example`.

James1912 replied:
Sure ill change them up

James1912 replied:
All references to `kubetest001` have been replaced with `couchdb001`

MHBauer commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,36 @@
+  name: kubetest001
+  namespace: default
+spec:
+  #clusterIP: 100.10.10.101
```
remove comment

kocolosk replied:
Remove this.

MHBauer commented on examples/storage/couchdb/README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Prerequisites
+Access to a kubernetes cluster running v1.4 or later, kubectl installed and
+configured.  Minikube was used for this example. Note that the service type
```
Can you write up a non-minikube example as an additional section? 

section one is:
- here is an example play deployment, where we use minikube
- let's poke at it
- let's clean it up

section two:
- for a prod environment, here is what changes, etc.

James1912 replied:
Sure thing :)

James1912 replied:
There is only really one difference for those who are not running minikube and thats how you get the external IP address of the service.  I have updated the README and mentioned how you go about getting the external IP of the service using `kubectl`

MHBauer commented on examples/storage/couchdb/README.md:
```diff
@@ -0,0 +1,84 @@
+every couchdb pod its own DNS name for its lifetime. 
+
+There are two services we need to create for our couchdb cluster.  First 
+theres the headless service for pod to pod discovery using DNS lookup and to 
```
theres -> there is OR there's

MHBauer reviewed (commented):
some more comments
@mikebrow @dug @kara as additional reviewers

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,91 @@
+    sys.exit(0)
+
+
+def discover_pod_fq_hostnames():
```
I'm worried that this function is brittle. What is it doing? Looking for SRV records associated with the FQDN of the Pod?

Is DNS preferable to querying the Kube API for this information? If so, does it make sense to use a library like http://www.dnspython.org/ ?

James1912 replied:
The petset docs recommend doing the service discovery using a DNS lookup so I just followed that.  I shall give that python module ago.

James1912 replied:
Replaced the use of nslookup with pydns and it's much cleaner now 😃

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,91 @@
+    system_dbs.pop(0)
+    try:
+        for system_db in system_dbs:
+            response = s.get(root_url.format(os.environ['POD_IP'], system_db))
```
Do you need to use the `POD_IP` here? Localhost should work fine.

James1912 replied:
Your right localhost would be fine here I shall switch it up.

kocolosk commented on examples/storage/couchdb/images/init_couch/configer_couch.sh:
```diff
@@ -0,0 +1,10 @@
+cp /core-config/vm.args $INSTALL_LOCATION/etc/vm.args
+mkdir $INSTALL_LOCATION/etc/local.d
+mkdir $INSTALL_LOCATION/etc/default.d
+sed -i $INSTALL_LOCATION/etc/vm.args -e "s/^\-name db/\-name db@$fqdn/"
```
This last one puzzles me; Erlang should default to using the FQDN when the node boots up. Is this shell script evaluated at a different time? I don't know why `hostname --fqdn` would work here but Erlang would fail to boot with `-name db`

kocolosk replied:
Also, what do you think about calling this `-name couchdb`?

James1912 replied:
This script is run before the the couchdb continer is created, it takes all of the files in the config map and pops them the etc folder.  Ow neat I did not know that Erlang would just use the FQDN if unset.  Sure thing I shall change up the name to couchdb

James1912 committed:
Renamed the Erlang node node to couchdb
Removed commented service IP
Removed all unnecessary port mapping
Renamed `foo-port` to `distributed-erlang-port`
Renamed ‘erlang-port’ to EPMD
Removed the unnecessary POD_IP use in the join_cluster script.
Fixed up some spelling in the README

James1912 committed:
Renamed all mentions of 'kubetest001' replacing it with `couchdb001`

James1912 committed:
Removed /./foo now that the entry points of the util continers are at '/'

James1912 committed:
Removed the nslookup cmd using pydns in its place.

James1912 commented:
@kocolosk @MHBauer Looking at the comments in the files changed tab.  I think I have now addressed all your concerns.

kocolosk commented:
This is improving, let me take another look and also put it through its paces.

kocolosk commented on examples/storage/couchdb/README.md:
```diff
@@ -0,0 +1,90 @@
+text password "changeme". On startup couchdb will hash the users password.  If 
+you would like to define your own admin users then just add them to local.ini 
+section of the core-config.  Also make sure that there is a valid admin user 
+credentials in the cluster admin secret at `core-config-maps/cluster-adm.yam`.
```
Definitely awkward that the admin credentials are written down in two different places, once in a ConfigMap and once in a Secret, and not synchronized. What would happen if we omitted them from the ConfigMap and instead had the init container create the admin account? Is that a bad idea?

kocolosk commented on examples/storage/couchdb/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## Couchdb 2.0.0 on Kubernetes
```
Let's be consistent and call it "CouchDB" everywhere - not "Couchdb" or "couchdb"

kocolosk commented on examples/storage/couchdb/README.md:
```diff
@@ -0,0 +1,90 @@
+[ConfigMap](http://kubernetes.io/docs/user-guide/configmap/) found at 
+`core-config-maps/core-configmap.yaml`. The `vm.args` file is where we set the
+erlang 'magic' cookie for all the couchdb instances alowing them to 
+communicate with each other.  Before we get to creating our couchdb cluster 
```
Can this magic cookie possibly be generated uniquely for each deployment?

kocolosk commented on examples/storage/couchdb/README.md:
````diff
@@ -0,0 +1,90 @@
+minikube service couchdb001 --url
+http://192.168.99.102:31391
+```
+For those not running minikube, you can discover the exsternal IP of the 
````
s/exsternal/external/

kocolosk commented on examples/storage/couchdb/README.md:
````diff
@@ -0,0 +1,90 @@
+```
+kubectl get service couchdb001  
+NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)                             AGE
+couchdb001   10.0.0.226   192.168.1.10  80/TCP,4369/TCP,9100/TCP,5986/TCP   16m
````
Suggest to expose the service on 5984 instead of 80 - that's our IANA registered port after all :)

kocolosk commented on examples/storage/couchdb/core-config-maps/core-configmap.yaml:
```diff
@@ -0,0 +1,493 @@
+
+    # Set the erlang cookie
+    -setcookie ZDc2ZjUzMWJkNTQyY2Zk
+  local.ini: |
```
Do we really need both .ini configuration files in here? Assuming the 2.0.0 image ships with the stock default.ini we could just have this local.ini override any defaults as needed.

kocolosk commented on examples/storage/couchdb/core-config-maps/core-configmap.yaml:
```diff
@@ -0,0 +1,493 @@
+    port = 5986
+
+    [couchdb]
+    uuid = 8b980cdf322417f7eefa5e2aaeb71563
```
You've got the UUID set here and in default.ini.

While we're on the topic ... this is one of those settings where we should try to start folks off on the right foot. In a clustered environment I believe the UUID is supposed to be shared by all nodes in a single cluster, and different clusters are supposed to have different UUIDs. Writing down a UUID in this file is important so that the nodes in the cluster share that identity, but should we go the extra step and configure a different one for each deployment?

This is basically the same discussion we could have for the Erlang magic cookie.

kocolosk commented on examples/storage/couchdb/images/init_couch/configure_couch.sh:
```diff
@@ -0,0 +1,10 @@
+cp /core-config/vm.args $INSTALL_LOCATION/etc/vm.args
+mkdir $INSTALL_LOCATION/etc/local.d
+mkdir $INSTALL_LOCATION/etc/default.d
+sed -i $INSTALL_LOCATION/etc/vm.args -e "s/^\-name couchdb/\-name couchdb@$fqdn/"
```
Is this line required? Typically Erlang will assume you want the FQDN if you use `-name` and leave the FQDN blank.

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,86 @@
+    hostnames = []
+    fq_pod_hostname = socket.getfqdn(os.environ['POD_IP'])
+    cluster_host = "{}.{}".format(fq_pod_hostname.split('.')[1],
+                                  fq_pod_hostname.split('.')[2])
```
As mentioned above, I think there's a more obvious way to extract the namespace and service name to use to make the DNS query.

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,86 @@
+    try:
+        dnslookup = resolver.query(cluster_host, "SRV")
+        for raw_hostname in dnslookup:
+            hostnames.append(str(raw_hostname)[8:-1])
```
The magic numbers are icky here. Also, isn't this supposed to return an instance of `dns.resolver.Answer`?

kocolosk replied:
Ah, I see what you're doing here. I'd suggest something like the following:

```
hostnames = [rdata.target.to_text() for rdata in dnslookup]
```

I suppose you can strip off the trailing dot if you want, but I'm wondering if it's necessary - that's the correct way to specify an absolute domain name in DNS.

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,86 @@
+
+    cluster_admin = (os.environ['CLUSTER_ADM_USER'].rstrip(), os.environ['CLUSTER_ADM_PASS'].rstrip())
+    populate_nodes_db(pod_hostnames, cluster_admin)
+    create_system_dbs(cluster_admin, system_dbs)
```
You really only want to execute `create_system_dbs` after the cluster is fully-connected, right? I'm worried about the possibility of creating conflicted system DBs here by executing this command for each Pod. Is there any way to defer it until all members of the initial Pet Set are up and running?

kocolosk replied:
Yep, confirmed this issue - the `_users` DB in my case has only 2 replicas even though I have a fully-connected cluster with 3 CouchDB nodes:

```
➜  couchdb git:(couchdb-petset) curl -s -u admin:changeme `minikube service couchdb001 --url`/_users/_shards | jq .
{
  "shards": {
    "00000000-1fffffff": [
      "couchdb@couchdb001-0.couchdb001-backend.default.svc.cluster.local",
      "couchdb@couchdb001-2.couchdb001-backend.default.svc.cluster.local"
    ],
    "20000000-3fffffff": [
      "couchdb@couchdb001-0.couchdb001-backend.default.svc.cluster.local",
      "couchdb@couchdb001-2.couchdb001-backend.default.svc.cluster.local"
    ],
    "40000000-5fffffff": [
      "couchdb@couchdb001-0.couchdb001-backend.default.svc.cluster.local",
      "couchdb@couchdb001-2.couchdb001-backend.default.svc.cluster.local"
    ],
...
```

kocolosk commented:
@James1912 for reasons I don't fully understand I ended up with a spurious cluster node after following these instructions:

```
➜  ~ curl -s -u admin:changeme `minikube service couchdb001 --url`/_membership | jq .
{
  "all_nodes": [
    "couchdb@couchdb001-0.couchdb001-backend.default.svc.cluster.local",
    "couchdb@couchdb001-1.couchdb001-backend.default.svc.cluster.local",
    "couchdb@couchdb001-2.couchdb001-backend.default.svc.cluster.local"
  ],
  "cluster_nodes": [
    "couchdb@ couchdb001-0.couchdb001-backend.default.svc.cluster.local",
    "couchdb@couchdb001-0.couchdb001-backend.default.svc.cluster.local",
    "couchdb@couchdb001-1.couchdb001-backend.default.svc.cluster.local",
    "couchdb@couchdb001-2.couchdb001-backend.default.svc.cluster.local"
  ]
}
```

kocolosk commented on examples/storage/couchdb/couchdb-petset-service.yaml:
```diff
@@ -0,0 +1,26 @@
+  ports:
+  - name: in-tcp-port
+    port: 80
+    targetPort: 5984
```
I was thinking through this element of the example further. This Service is not necessary to create the CouchDB cluster, right? It's only used to expose the cluster to consumers. In that case I would not expose EPMD, the Erlang distribution port, or the node-local port here. Those ports are intended to be specific to each Pod; anyone who needs to access them should be accessing them via the Pod's IP and not via this Service. I would remove all of those, and just expose `port: 5984`.

kocolosk commented on examples/storage/couchdb/core-config-maps/core-configmap.yaml:
```diff
@@ -0,0 +1,493 @@
+
+    ; Custom settings should be made in this file. They will override settings
+    ; in default.ini, but unlike changes made to default.ini, this file won't be
+    ; overwritten on server upgrade.
```
The 3 lines above certainly don't belong in this file.

kocolosk commented on examples/storage/couchdb/images/readiness_probe/join_cluster.py:
```diff
@@ -0,0 +1,86 @@
+            if node_doc not in existing_hostnames:
+                r = s.put("{0}/couchdb%40{1}".format(node_db_url, pod_hostname),
+                          data='{}')
+                return r
```
`return` is the wrong keyword here. This will cause you to exit prematurely from the for loop and you'll only execute one PUT request. I tried this manually and found that I had to run `join_cluster` twice to establish a fully-connected cluster.

Best to just remove this line.

thaume commented on examples/storage/couchdb/couchdb-petset.yaml:
```diff
@@ -0,0 +1,108 @@
+            {
+            "name": "configer-couch",
+            "image": "glyndwr/kube-init-couchdb",
+            "command": ["/bin/bash", "-c", "./configer_couch.sh"],
```
I think it should be "configure_couch.sh" here right ?

kocolosk replied:
Yes it looks like @James1912 corrected the spelling in 5c68009 and forgot to update it here. Good catch.

thaume commented on examples/storage/couchdb/images/init_couch/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM debian:jessie
+
+ADD configer_couch.sh /
+RUN chmod +x configer_couch.sh
```
Shouldn't it be "configure_couch.sh" here as well ?

kocolosk replied:
Ditto

thaume commented:
Hey there,

I gave a shot to your example and it worked great for me. The only thing is that I had to create the DBs by hand. Maybe it could be nice to have a script automatically create the base DBs ? (_nodes, _replications, _users, etc...).

What do you think ?
